[BOLT][AArch64] Implement PLTCall optimization

paschalis-mpeis opened the pull request:
`convertCallToIndirectCall` applies the PLTCall optimization and returns an (updated if needed) iterator to the converted call instruction. Since AArch64 requires to inject additional instructions to implement this pass, the relevant BasicBlock and an iterator was passed to the `convertCallToIndirectCall`.

`NumCallsOptimized` is updated only on successful application of the pass.

Tests:
- Inputs/plt-tailcall.c: an example of a tail call optimized PLT call.
- AArch64/plt-call.test: it is the actual A64 test, that runs the PLTCall optimization on the above input file and verifies the application of the pass to the calls: 'printf' and 'puts'.

paschalis-mpeis committed:
[BOLT][AArch64] Implement PLTCall optimization

`convertCallToIndirectCall` applies the PLTCall optimization and returns
an (updated if needed) iterator to the converted call instruction.
Since AArch64 requires to inject additional instructions to implement this
pass, the relevant BasicBlock and an iterator was passed to the
`convertCallToIndirectCall`.

`NumCallsOptimized` is updated only on successful application of the pass.

Tests:
- Inputs/plt-tailcall.c: an example of a tail call optimized PLT call.
- AArch64/plt-call.test: it is the actual A64 test, that runs the PLTCall
  optimization on the above input file and verifies the application of
  the pass to the calls: 'printf' and 'puts'.

paschalis-mpeis commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -1412,9 +1413,15 @@ class MCPlusBuilder {
+  /// call instruction.
+  ///
+  /// \return true on success
+  virtual bool convertCallToIndirectCall(BinaryBasicBlock &BB,
```
On Aarch64, `convertCallToIndirectCall` does more than a direct->indirect call conversion to implement this optimization. If this is considered misleading we could rename the method.

paschalis-mpeis committed:
Refactored convertCallToIndirectCall to createIndirectPltCall

createIndirectPltCall now returns a list of instructions. All
BinaryBasicBlock-related logic is now on PLTCall pass.

Added the same test for x86, to ensure nothing breaks on that backend.

rafaelauler commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -1412,9 +1413,15 @@ class MCPlusBuilder {
+  /// call instruction.
+  ///
+  /// \return true on success
+  virtual bool convertCallToIndirectCall(BinaryBasicBlock &BB,
```
Hi Paschalis, 

Thanks a lot for submitting this contribution to the project, I appreciate it!

One thing that might not be obvious in the code base is that the MCPlusBuilder layer (or the BOLT targets lib, for that matter) do not depend on BOLT IR classes, such as BinaryBasicBlock. They are limited to use MCInst and the class MCPlusBuilder::InstructionIterator whenever the API needs to communicate a walkable MCInst pointer to target-specific code.

If you add this dependency of bolt target libs on bolt IR classes, this will create a cyclic dependency that will likely break our shared lib build (if you build BOLT with enable_shared you might see this). I would suggest you taking a look at other methods in X86MCPlusBuilder to see how we implement this kind of optimization -- usually by moving code that depends on BB knowledge to the pass itself, instead of doing that logic in target libs.

samolisov replied:
@paschalis-mpeis There is a polymorphic iterator `InstructionIterator`, which is defined in `MCPlusBuilder.h` and there is a pattern to pass every required instruction sequence to `MCPlusBuilder`'s methods by this iterator (but, as I see, it is used to pass a chain of instruction with no modifications of the iterator itself). Also there is the `InstructionListType` and, please have a look at the [`createLoadImmediate`](https://github.com/llvm/llvm-project/blob/main/bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp#L1453) for example, the generated in `MCPlusBuilder` instruction list should be returned by value.

paschalis-mpeis replied:
Thanks @rafaelauler  and @samolisov  for the suggestions. I think the way to go is using `InstructionListType`.

I will rename `convertCallToIndirectCall` to `createIndirectPltCall`, addressing my previous concern [here](https://github.com/llvm/llvm-project/pull/93584#discussion_r1617667657), but feel free to suggest a better name.
The function will be adjusted to return an `InstructionListType`, and anything that relates to Binary BB logic will be moved to the pass itself.

I think the closest existing example of this is with the `InlineMemcpy` pass and `createInlineMemcpy`.

samolisov commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -1055,6 +1055,52 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+                          /* Addend */ 0, Ctx, ELF::R_AARCH64_LD64_GOT_LO12_NC);
+    It = BB.insertInstruction(It, InstLoad);
+
+    MCInst InstAdrp;
```
Why the `LDR` was generated before the corresponding `ADRP` even though it is inserted after? If I get it correctly, in the result code the following snippet will be generated:

```
adrp reg1, <symbol>
ldr  reg2, [reg1, <offset>]
```

I believe the code is more readable when the instruction is generated in the code one by one to reflect the result without playing with insertion points (`It`) in this case. Also, this let us to increment `It` after every instruction without using a magic constant (`2` in `It = It + 2`). Is there a dependency between `InstLoad` and `InstAdrp` what makes the reflection of the generated code impossible?

paschalis-mpeis replied:
Correct, thanks @yota9 . The generated code is shown as a comment on the top.

@samolisov I agree that it'll look more natural. The adjusted function that will be returning a `InstructionListType` would allow this.

paschalis-mpeis commented on bolt/lib/Passes/PLTCall.cpp:
```diff
@@ -61,19 +61,23 @@ Error PLTCall::runOnFunctions(BinaryContext &BC) {
-                                          BC.Ctx.get());
-        BC.MIB->addAnnotation(Instr, "PLTCall", true);
-        ++NumCallsOptimized;
+        if (BC.MIB->convertCallToIndirectCall(BB, It, CalleeBF->getPLTSymbol(),
```
That is true. The method now returns an InstructionListType, so this was adjusted accordingly.

samolisov commented on bolt/lib/Passes/PLTCall.cpp:
```diff
@@ -61,18 +61,20 @@ Error PLTCall::runOnFunctions(BinaryContext &BC) {
-        BC.MIB->addAnnotation(Instr, "PLTCall", true);
+        const InstructionListType NewCode = BC.MIB->createIndirectPltCall(
+            *II, CalleeBF->getPLTSymbol(), BC.Ctx.get());
+        II = BB.replaceInstruction(II, NewCode);
```
`std::advance()` after this line uses `NewCode.size() - 1` as the second argument. I think it may make sense to ensure that the size is not zero adding a corresponding `assert`.

paschalis-mpeis replied:
Thanks for the suggestion, added.

samolisov commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -1055,6 +1055,47 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+
+    return Code;
+  }
+
```
This function looks good for me now.

paschalis-mpeis committed:
Addressing reviewers

paschalis-mpeis commented:
Rebased to latest main and addressed latest review.

---

NOTE: some unrelated commits on main were causing failures on unrelated BOLT tests.
Now that tests seem OK, I was able to do the rebase.

Those commits were:
- Errors introduced by:
  - 435ea21c897f
  - 20d497c26fc
- Fixed by:
  - 0353f6a
  - 6ef632a
  - #94301

yota9 reviewed (approved):
LGTM Thank you! Give 1-2 days for other reviewers to evaluate change please

paschalis-mpeis commented:
Hi all,

if there are not further comments, I will be merging this PR by EOD.
Once again thank you all for the feedback and suggestions.

MaskRay commented:
X86/plt-call.test and AArch64/plt-call.test test PLT handling . They should link against a dummy DSO to not rely on an unguaranteed linker behavior.  Sent #125625
